pkg/rbac/clusterScopedAuthorization.go

```diff
@@ -0,0 +1,78 @@
+package rbac
+
+import (
+	"context"
+	"time"
+
+	"github.com/doug-martin/goqu/v9"
+	"github.com/stolostron/search-v2-api/pkg/config"
+	"k8s.io/klog/v2"
+)
+
+type sharedList struct {
+	err       error
+	resources map[string][]string
+	updatedAt time.Time
+}
+
+func (cache *Cache) checkUserResources() (sharedList, error) {
+
+	cache.sharedLock.Lock()
+	//check if we already have the resources in cache
+	cr := cache.shared
+	if cr.resources != nil && time.Now().Before(cr.updatedAt.Add(time.Duration(config.Cfg.SharedCacheTTL)*time.Second)) {
+		klog.V(5).Info("Using cluster scoped resources from cache.")
+		cache.sharedLock.Unlock()
+		return cr, nil
+	}
+	cache.sharedLock.Unlock()
+
+	if cr.resources != nil && cr.err != nil {
+		return cr, cr.err
+	}
+
+	//otherwise build query and get cluster-scoped resources for user from database and cache:
+	err := cache.getClusterScopedResources()
+
+	return cr, err
+}
+
+func (cache *Cache) getClusterScopedResources() error {
+
+	schemaTable := goqu.S("search").Table("resources")
+	ds := goqu.From(schemaTable)
+
+	//"SELECT DISTINCT(data->>apigroup, data->>kind) FROM search.resources WHERE
+	// cluster='local-cluster' AND namespace=NULL"
+	query, _, err := ds.SelectDistinct(goqu.COALESCE(goqu.L(`"data"->>'apigroup'`), "").As("apigroup"),
+		goqu.COALESCE(goqu.L(`"data"->>'kind'`), "").As("kind")).
+		Where(goqu.L(`"cluster"::TEXT = 'local-cluster'`), goqu.L(`"data"->>'namespace'`).IsNull()).ToSQL()
+	if err != nil {
+		klog.Errorf("Error creating query [%s]. Error: [%+v]", query, err)
+	}
+	rows, queryerr := cache.pool.Query(context.Background(), query)
+	if queryerr != nil {
+		klog.Errorf("Error resolving query [%s]. Error: [%+v]", query, queryerr.Error())
+	}
+	if rows != nil {
+		defer rows.Close()
+		csrmap := make(map[string][]string)
+
+		for rows.Next() {
+			var kind string
+			var apigroup string
+			err := rows.Scan(&apigroup, &kind)
+			if err != nil {
+				klog.Errorf("Error %s retrieving rows for query:%s for apigroup %s and kind %s", err.Error(), query, apigroup, kind)
+			}
```

Comment:
Errors are per row, so the result could have 3 rows and only 1 has an error.
If a row has an error we should skip adding it and save the error.  
```suggestion
			err := rows.Scan(&apigroup, &kind)
			if err != nil {
				klog.Errorf("Error %s retrieving rows for query:%s for apigroup %s and kind %s", err.Error(), query, apigroup, kind)
				sharedList.err = err // This will only keep the last row with an error. It's unlikely that we see an error here, so reporting the last one is "good enough". The log above will record all instances.
				continue // skip adding this row because it had errors.
			}
```